### PR TITLE
[MPS] Honor include_last_offset and scale_grad_by_freq in embedding_bag

### LINKED ISSUE #190060: [MPS] embedding_bag(include_last_offset=True) ignores the terminal offset
### 🐛 Describe the bug

With `include_last_offset=True`, the final entry of `offsets` must end the
last bag. The MPS host decrements `num_bags` accordingly, but the kernel
always ends the final bag at `num_indices`, so the supplied terminal offset is
never read and trailing indices are wrongly summed into the last bag.

```python
import torch

W = torch.arange(16.0).reshape(4, 4)
idx = torch.tensor([0, 1, 2])
offs = torch.tensor([0, 2])  # one bag: indices [0, 2)
ref = torch.nn.functional.embedding_bag(idx, W, offs, mode="sum", include_last_offset=True)
out = torch.nn.functional.embedding_bag(idx.to("mps"), W.to("mps"), offs.to("mps"),
                                        mode="sum", include_last_offset=True)
print("cpu:", ref.tolist())
print("mps:", out.cpu().tolist())
```

Output on a source build of current main (2.14.0a0+git2662808); also
reproduces on 2.13.0 stable:

```text
cpu: [[4.0, 6.0, 8.0, 10.0]]
mps: [[12.0, 15.0, 18.0, 21.0]]   # row 2 wrongly included
```

The host decrement is at
https://github.com/pytorch/pytorch/blob/26628087023e5102381849e9a27c3f09b6d485a4/aten/src/ATen/native/mps/operations/EmbeddingBag.mm#L70-L73
and the kernel's last-bag end at
https://github.com/pytorch/pytorch/blob/26628087023e5102381849e9a27c3f09b6d485a4/aten/src/ATen/native/mps/kernels/EmbeddingBag.metal#L172-L176
uses `num_indices` unconditionally. Backward gradients for the trailing
indices are correspondingly wrong. #52851 documents the expected
`include_last_offset` contract generally (it predates the MPS backend).

### Versions

PyTorch 2.14.0a0+git2662808 (source build of main) and 2.13.0; macOS 26.5.2 (arm64), Apple M5 Max, Python 3.12.13.

cc @ezyang @gchanan @kadeng @msaroufim @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

### LINKED ISSUE #190061: [MPS] embedding_bag(scale_grad_by_freq=True) silently ignores the flag
### 🐛 Describe the bug

`scale_grad_by_freq=True` must divide each row's weight gradient by the
frequency of its index. The flag reaches
`_embedding_bag_dense_backward_mps` but is never used, so MPS silently
returns the unscaled gradient.

```python
import torch

for dev in ("cpu", "mps"):
    w = torch.randn(4, 3).to(dev).requires_grad_()
    idx = torch.tensor([1, 1], device=dev)
    offs = torch.tensor([0], device=dev)
    out = torch.nn.functional.embedding_bag(idx, w, offs, mode="sum", scale_grad_by_freq=True)
    out.sum().backward()
    print(dev, "row-1 grad:", w.grad[1].tolist())
```

Output on a source build of current main (2.14.0a0+git2662808); also
reproduces on 2.13.0 stable:

```text
cpu row-1 grad: [1.0, 1.0, 1.0]
mps row-1 grad: [2.0, 2.0, 2.0]
```

The flag arrives at
https://github.com/pytorch/pytorch/blob/26628087023e5102381849e9a27c3f09b6d485a4/aten/src/ATen/native/mps/operations/EmbeddingBag.mm#L185-L194
and is unused in the params or kernel dispatch. #185119 fixed the identical
silently-ignored flag for `nn.Embedding` on MPS; `nn.EmbeddingBag` was left
out.

### Versions

PyTorch 2.14.0a0+git2662808 (source build of main) and 2.13.0; macOS 26.5.2 (arm64), Apple M5 Max, Python 3.12.13.

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

### PR DESCRIPTION
**Pull request by bogatyy**
**include_last_offset** (#190060): the host decremented `num_bags` but the kernel always ended the final bag at `num_indices`, so the supplied terminal offset was never read and trailing indices were wrongly folded into the last bag, in forward and backward. The kernel now ends the last bag at `offsets[num_bags]` when the flag is set. Trailing indices belong to no bag, so forward marks their `offset2bag` entries with `-1` (previously they'd have been uninitialized memory) and the backward kernels skip them.

**scale_grad_by_freq** (#190061): the flag reached `_embedding_bag_dense_backward_mps` but was never used. Each row's gradient is now divided by its index's frequency, per the documented semantics.

Two CPU quirks surfaced while writing the tests, so CPU parity is asserted only where CPU is correct:
- CPU's dense backward indexes its counts array with the unique-group counter instead of the index value (`counts[indices_data[i]]` at [EmbeddingBag.cpp:1573](https://github.com/pytorch/pytorch/blob/26628087023e5102381849e9a27c3f09b6d485a4/aten/src/ATen/native/EmbeddingBag.cpp#L1573)), so a once-occurring index that sorts after a repeated one gets scaled by the wrong count. Mixed-count cases are tested against hand-computed values; I'll file this separately.
- With a trailing index (terminal offset < len(indices)), CPU excludes it from sum but leaks it into mean's divisor and max's candidates, and gives it a nonzero per-sample-weights gradient (#52851 territory). MPS now implements the CSR contract consistently: trailing indices participate in nothing.

Coverage: terminal == len(indices) CPU parity for sum/mean/max forward+backward, trailing-index sum (CPU parity) and mean/max (contract values), per_sample_weights with a trailing index, frequency scaling with uniform and mixed counts. All 56 embedding tests pass.

Related: #187572 (offsets validation) touches the same kernel; the diffs shouldn't conflict beyond proximity.

Fixes #190060
Fixes #190061

*This PR was authored with assistance from Claude.*

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/EmbeddingBag.mm [hunk 1]
```diff
@@ -247,6 +255,13 @@ Tensor _embedding_bag_dense_backward_mps(const Tensor& output_grad,
     }
   });
 
+  if (scale_grad_by_freq) {
+    // Match CPU/CUDA: divide each row's gradient by its index's frequency.
+    auto counts = at::zeros({num_weights}, output_grad.options());
+    counts.index_add_(0, indices, at::ones_like(indices, output_grad.options()));
+    weight_grad.div_(counts.clamp_min(1).unsqueeze(1));
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/EmbeddingBag.mm]:
```suggestion
    auto counts = at::bincount(indices, {}, num_weights)
                  .clamp_min_(1)
                  .to(output_grad.scalar_type());
    
    weight_grad.div_(counts.unsqueeze(1));

**Review by Skylion007 — COMMENTED** [review]
This might end up resolving to the same thing TBH, not sure.

File: aten/src/ATen/native/mps/operations/EmbeddingBag.mm [hunk 2]
```diff
@@ -18,6 +18,9 @@
 #include <ATen/ops/_embedding_bag_native.h>
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/EmbeddingBag.mm:18]: also add the bincount header here

File: aten/src/ATen/native/mps/operations/EmbeddingBag.mm (hunk 1 above)
Comment (bogatyy) [reply aten/src/ATen/native/mps/operations/EmbeddingBag.mm]: Applied in 3fe521fe, thanks. Your hunch that it resolves to the same thing turned out to be right numerically: index_add_ on MPS upcasts bf16 accumulation internally, so both versions produce identical counts. But bincount is still strictly better, one dispatch instead of three and int64 accumulation guaranteed by construction rather than by an implementation detail of index_add_. Also added a bf16 case to the test that pins the exact divisor for an index appearing 300 times (past bf16's 256 integer limit), so a future regression to naive reduced-precision counting would be caught.

File: aten/src/ATen/native/mps/operations/EmbeddingBag.mm (hunk 2 above)
Comment (bogatyy) [reply aten/src/ATen/native/mps/operations/EmbeddingBag.mm:18]: Added in 3fe521fe (and dropped the now-unused ones_like.h).

**Comment by bogatyy** [comment]
@malfet Rebased onto current trunk and moved the test strategy to the device-generic suite as you suggested: test/nn/test_embedding.py now instantiates for MPS, picking up the existing embedding_bag coverage including the mixed-counts scale_grad_by_freq test from #190264. float64 combos are excluded via dtypesIfMPS, sparse-grad and double-gradcheck tests are skipped, and the out-of-bounds test is skipped since MPS reports bounds errors asynchronously. The test_mps.py additions now cover only the MPS-specific parts: trailing-index CSR semantics under include_last_offset (where CPU is inconsistent, #52851) and bf16 count-divisor exactness. Enabling the generic suite also exposed a pre-existing bug, fixed here: the per_sample_weights backward dispatched its kernel on the saved unpromoted indices dtype while offset2bag arrives promoted, so int32 indices with int64 offsets scaled psw gradients by the wrong bags.

To verify the coverage split, I ran the enabled generic suite against stock kernels (this PR's test changes, main's aten code): the mixed-counts test catches the ignored scale_grad_by_freq, and six int32/int64 per-sample-weights combos catch the promotion bug, but every generic include_last_offset case passes on stock. That is because stock already sized the output correctly and only bounded the last bag by num_indices instead of offsets[num_bags], and those coincide for well-formed inputs, which is all a shared test can use given CPU's trailing-index inconsistency. The retained test_mps.py trailing-index test is the only regression coverage for that part of the fix.

**Comment by malfet** [comment]
@bogatyy Why do you still need test_mps.py changes? Aren't tests in test_embeddings sufficient? I.e. if CPU is doing something wrong, why not fix CPU first or create a test and mark it as `xfailIfCPU`? 

I.e. `test_mps.py` should eventually only contain regressions which are unique to the framework, but majority of the test shoudl be accelerator (CUDA, ROCM, MPS, ....) agnostic....

**Comment by bogatyy** [comment]
@malfet CPU does not implement strict CSR, and turns out CUDA does not either.
I'll move the test to `test_embedding.py`, mark expectedFailureCPU + expectedFailureCUDA, and it can be fixed from there.

**Comment by bogatyy** [comment]
Done in 8b9b9e8, test_mps.py now has zero embedding changes. The trailing-index test asserts strict CSR device-generically with expectedFailureCPU (#52851) and expectedFailureCUDA (#29019, the EmbeddingBag.cu source notes include_last_offset still routes to the legacy last-bag bound). The expectedFailure decorators raise on unexpected success, so fixing a backend automatically prompts removing its marker. The bf16 exact-counts test moved too since CPU and CUDA count in integer math and pass it. Both are gated to cpu/cuda/mps as I could not verify XPU behavior.

**Comment by bogatyy** [comment]
Rebased over #187572, which touched the same last-bag boundary code. The resolution composes the two: this PR's include_last_offset-aware indices_end is now checked by that PR's in-kernel span validation, so an invalid terminal offset reports through the error buffer instead of clamping silently. All embedding tests pass on a source build, including the offsets-validation tests that landed with #187572.
